File: Golang/06. Software Design Principles SOLID & More/practice-19.go
/**
 * Practice 19: Applying DRY (Don't Repeat Yourself)
 * Task: Refactor duplicated discount and invoice logic into reusable components.
 *
 * How to run:
 *   go run practice-19.go
 */

package main

import "fmt"

type OrderItem struct {
	Name      string
	UnitPrice float64
	Quantity  int
}

func (i *OrderItem) LineTotal() float64 {
	return i.UnitPrice * float64(i.Quantity)
}

type DiscountPolicy struct{}

func (d *DiscountPolicy) Calculate(subtotal float64, customerType string) float64 {
	rate := 0.0
	switch customerType {
	case "gold":
		rate = 0.10
	case "silver":
		rate = 0.05
	default:
		rate = 0.00
	}
	return subtotal * rate
}

type InvoiceSummary struct {
	Subtotal float64
	Discount float64
	Vat      float64
	Total    float64
}

type InvoiceCalculator struct {
	DiscountPolicy *DiscountPolicy
}

func (c *InvoiceCalculator) Summarize(items []OrderItem, customerType string) InvoiceSummary {
	subtotal := 0.0
	for _, item := range items {
		subtotal += item.LineTotal()
	}

	discount := c.DiscountPolicy.Calculate(subtotal, customerType)
	taxable := subtotal - discount
	vat := taxable * 0.05

	return InvoiceSummary{
		Subtotal: subtotal,
		Discount: discount,
		Vat:      vat,
		Total:    taxable + vat,
	}
}

func printInvoice(customerName string, customerType string, items []OrderItem) {
	calculator := &InvoiceCalculator{DiscountPolicy: &DiscountPolicy{}}
	summary := calculator.Summarize(items, customerType)

	fmt.Printf("Customer: %s (%s)\n", customerName, customerType)
	for _, item := range items {
		fmt.Printf("  - %-10s x%-2d @ $%6.2f = $%7.2f\n",
			item.Name, item.Quantity, item.UnitPrice, item.LineTotal())
	}

	fmt.Printf("Subtotal : $%.2f\n", summary.Subtotal)
	fmt.Printf("Discount : $%.2f\n", summary.Discount)
	fmt.Printf("VAT (5%%) : $%.2f\n", summary.Vat)
	fmt.Printf("Total    : $%.2f\n", summary.Total)
	fmt.Println()
}

func main() {
	items := []OrderItem{
		{Name: "Keyboard", UnitPrice: 45.0, Quantity: 2},
		{Name: "Mouse", UnitPrice: 20.0, Quantity: 1},
		{Name: "USBC", UnitPrice: 12.5, Quantity: 3},
	}

	printInvoice("Samia", "regular", items)
	printInvoice("Afsana", "silver", items)
	printInvoice("Hasan", "gold", items)
}

File: C#/06. Software Design Principles SOLID & More/Practice19.cs
/**
 * Practice 19: Applying DRY (Don't Repeat Yourself)
 * Task: Refactor duplicated discount and invoice logic into reusable components.
 *
 * How to run (using dotnet-script):
 *   dotnet script Practice19.cs
 */

using System;
using System.Collections.Generic;

class Practice19
{
    class OrderItem
    {
        public string Name;
        public decimal UnitPrice;
        public int Quantity;

        public OrderItem(string name, decimal unitPrice, int quantity)
        {
            Name = name;
            UnitPrice = unitPrice;
            Quantity = quantity;
        }

        public decimal LineTotal => UnitPrice * Quantity;
    }

    static class DiscountPolicy
    {
        public static decimal Calculate(decimal subtotal, string customerType)
        {
            var key = customerType.ToLowerInvariant();
            decimal rate = key == "gold" ? 0.10m : key == "silver" ? 0.05m : 0.00m;
            return subtotal * rate;
        }
    }

    class InvoiceSummary
    {
        public decimal Subtotal;
        public decimal Discount;
        public decimal Vat;
        public decimal Total;
    }

    class InvoiceCalculator
    {
        private const decimal VatRate = 0.05m;

        public InvoiceSummary Summarize(List<OrderItem> items, string customerType)
        {
            decimal subtotal = 0m;
            foreach (var item in items)
                subtotal += item.LineTotal;

            decimal discount = DiscountPolicy.Calculate(subtotal, customerType);
            decimal taxable = subtotal - discount;
            decimal vat = taxable * VatRate;

            return new InvoiceSummary
            {
                Subtotal = subtotal,
                Discount = discount,
                Vat = vat,
                Total = taxable + vat
            };
        }
    }

    static void PrintInvoice(string customerName, string customerType, List<OrderItem> items)
    {
        var calculator = new InvoiceCalculator();
        var summary = calculator.Summarize(items, customerType);

        Console.WriteLine($"Customer: {customerName} ({customerType})");
        foreach (var item in items)
            Console.WriteLine($"  - {item.Name,-10} x{item.Quantity,-2} @ ${item.UnitPrice,6:F2} = ${item.LineTotal,7:F2}");

        Console.WriteLine($"Subtotal : ${summary.Subtotal:F2}");
        Console.WriteLine($"Discount : ${summary.Discount:F2}");
        Console.WriteLine($"VAT (5%) : ${summary.Vat:F2}");
        Console.WriteLine($"Total    : ${summary.Total:F2}");
        Console.WriteLine();
    }

    static void Main(string[] args)
    {
        var items = new List<OrderItem>
        {
            new OrderItem("Keyboard", 45.0m, 2),
            new OrderItem("Mouse", 20.0m, 1),
            new OrderItem("USBC", 12.5m, 3)
        };

        PrintInvoice("Samia", "regular", items);
        PrintInvoice("Afsana", "silver", items);
        PrintInvoice("Hasan", "gold", items);
    }
}
